consensus/src/transaction_shuffler/use_case_aware/delayed_queue.rs

```diff
@@ -0,0 +1,503 @@
+                            vacant.insert(use_case);
+                        },
+                    }
+                } else {
+                    // Account tracked and not empty, so appending a new txn to it won't affect positions
+                    // in delay queues
+                    account.queue_txn(input_idx, txn);
+                }
+            },
+            None => {
+                // Account not previously tracked.
+                let account = Account::new_with_txn(self.output_idx + 1, input_idx, txn);
+                // Account didn't exist before, so use case must have been tracked, otherwise the
+                // txn whould've been selected for output, bypassing the queue.
+                let use_case = self.use_cases.expect_mut(&use_case_key);
+                if use_case.is_empty() {
+                    self.use_case_placeholders_by_delay
+                        .strict_remove(&use_case.delay_key());
+                } else {
+                    self.use_cases_by_delay.strict_remove(&use_case.delay_key());
+                }
+                use_case.add_account(address, &account);
+
+                self.accounts.strict_insert(address, account);
+                self.use_cases_by_delay
+                    .strict_insert(use_case.delay_key(), use_case_key.clone());
+            },
+        }
+    }
+
+    /// Txn from input queue directly selected for output, needs to bump delays for relevant
+    /// account and use case.
+    fn update_delays_for_selected_txn(
+        &mut self,
+        input_idx: InputIdx,
+        address: AccountAddress,
+        use_case_key: UseCaseKey,
+    ) {
+        let account_try_delay_till = self.output_idx + 1 + self.config.sender_spread_factor();
+        let use_case_try_delay_till =
+            self.output_idx + 1 + self.config.use_case_spread_factor(&use_case_key);
+
+        match self.use_cases.entry(use_case_key.clone()) {
+            hash_map::Entry::Occupied(occupied) => {
+                let use_case = occupied.into_mut();
+                // Txn wouldn't have been selected for output if the use case is empty (tracking
+                // for a try_delay_till > self.output_idx)
+                assert!(!use_case.is_empty());
+
+                self.use_cases_by_delay.strict_remove(&use_case.delay_key());
+                use_case.update_try_delay_till(use_case_try_delay_till);
+                self.use_cases_by_delay
+                    .strict_insert(use_case.delay_key(), use_case_key);
+            },
+            hash_map::Entry::Vacant(vacant) => {
+                let use_case = UseCase::new_empty(use_case_try_delay_till, input_idx);
+                self.use_case_placeholders_by_delay
+                    .strict_insert(use_case.delay_key(), use_case_key);
+                vacant.insert(use_case);
+            },
+        }
+
+        // Account must not have been tracked before, otherwise the txn wouldn't have been selected
```

Comment:
Read the comment in context: the function is `update_delays_for_selected_txn()` which is called by `queue_or_return()`.. does the "otherwise" make sense now?

Reply:
Thanks for the explaination, this makes sense but the code structure is still a bit confusing. How about breaking this into two functions `update_usecase_delay_for_selected_txn` and `add_account_with_delay` something like that and call it from `queue_or_return`.  Otherwise, the account being not tracked here is not super obvious as the context is in `queue_or_return` function.